compiler/transforms/acc_cse.py

```diff
@@ -53,15 +54,90 @@ def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
         )
 
 
-class AccCse(ModulePass):
+class HoistSetupCallsIntoConditionals(RewritePattern):
     """
-    Common subexpression elimination for the `acc` (accelerator) dialect.
+    Hoists setup calls into scf.if blocks.
 
-    This pass rewrites acc dialect operations to find and simplify redundant
-    setup calls.
+    This will never result in additional setup calls and only
+    reduced the number of fields that are set up.
+    (proof left as an exercise to the reader)
+
+    Things we need to worry about:
+
+    - not inserting a setup op before the originally produced state
+      is launched.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: acc.SetupOp, rewriter: PatternRewriter, /):
+        # do not apply if our in_state is not an scf.if
+        if op.in_state is None or not isinstance(op.in_state.owner, scf.If):
+            return
+        # grab some helper vars
+        old_in_state = op.in_state
+        assert isinstance(old_in_state, OpResult)
+
+        # Step 1: Check that it's legal to move:
+        uses = tuple(
+            use for use in op.in_state.uses if isinstance(use.operation, acc.LaunchOp)
+        )
+        # if we have some launch ops, we need to investigate further:
+        if len(uses) > 0:
+            # if there is a launch outside the scf.if
+            launch_op = uses[0].operation
+            # check that it is in the same block (if not bail out)
+            # TODO: better check this case?
+            if launch_op.parent_block() is not op.parent_block():
+                return
```

Comment:
```mlir
scf.if %b  {
  acc.launch
} else {
  ...
}
acc.setup
```

Can you not still lift the setup op in the if region? I think you can make an exception if the `launch_op.parent_block()` is part of the `scf.if` under consideration

Reply:
I think you missed one tiny detail, the IR would look like this:

```mlir
%state_before_if = acc.setup()
acc.launnch(%state_before_if)
%state_after_if = scf.if (%i1) -> !acc.state {
  %s1 = acc.setup(%state_before_if)
  acc.launch(%s1)
  scf.yield %s1
} else {
  scf.yield %state_before_if
}

%s_new = acc.setup(%state_after_if) // <- this is our `op`
acc.launch(%s_new)
```

Here we are looking at the `%s_new = acc.setup(%state_after_if)` as our `op` variable.

The input state is the variable `%state_after_if`, and we don't find the `acc.launch(%s1)` op as a use of it.

The only way a launch op use of the `%state_after_if` cannot share a block with our op is if it is inside another nested op *after* the `if`:

```mlir
%state_before_if = acc.setup()
acc.launnch(%state_before_if)
%state_after_if = scf.if (%i1) -> !acc.state {
// ...

// either here (illegal):
some.op() {
  acc.launch(%state_after_if)
}

%s_new = acc.setup(%state_after_if) // <- this is our `op`
acc.launch(%s_new)



// or here (legal):
some.op() {
  acc.launch(%state_after_if)
}
```

But accounting for this structure is not in scope for this PR

Reply:
ah, yes, this makes a lot of sense. thank you for the elaboration!